Source/Core/Common/BlockingLoop.h

```diff
@@ -0,0 +1,164 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <atomic>
+#include <mutex>
+#include <thread>
+
+#include "Common/Event.h"
+#include "Common/Flag.h"
+
+namespace Common
+{
+
+// This class provides a synchronized loop.
+// It's a thread-safe way to trigger a new iteration without busy loops.
+// It's optimized for high-usage iterations which usually are already running while it's triggered often.
+class BlockingLoop
+{
+public:
+	BlockingLoop()
+	{
+		m_running.Clear();
+	}
+
+	~BlockingLoop()
+	{
+		Stop();
+	}
+
+	// Triggers to rerun the payload of the Run() function at least once again.
+	// This function will never block and is designed to finish as fast as possible.
+	void Set()
+	{
+		// already running, so no need for a wakeup
+		if (m_is_running.IsSet())
+			return;
+
+		m_is_running.Set();
+		m_is_pending.Set();
+		m_new_work_event.Set();
+	}
+
+	// Wait for a complete payload run after the last Set() call.
+	// If stopped, this returns immediately.
+	void Wait()
+	{
+		// We have to give the loop a chance to exit.
+		m_may_sleep.Set();
+
+		if (!m_running.IsSet() || (!m_is_running.IsSet() && !m_is_pending.IsSet()))
+			return;
+
+		// notifying this event will only wake up one thread, so use a mutex here to
+		// allow only one waiting thread. And in this way, we get an event free wakeup
+		// but for the first thread for free
+		std::lock_guard<std::mutex> lk(m_wait_lock);
+
+		while (m_running.IsSet() && (m_is_running.IsSet() || m_is_pending.IsSet()))
+		{
+			m_may_sleep.Set();
+			m_done_event.Wait();
+		}
+	}
+
+	// Half start the worker.
+	// So this object is in running state and Wait() will block until the worker calls Run().
+	// This may be called from any thread and is supposed to call at least once before Wait() is used.
+	void Prepare()
+	{
+		// There is a race condition if the other threads is still in calling this function
+		// while the loop thread is initializing. Using this lock will ensure a valid state.
```

Comment:
I think this should be, "... if the other threads call this function while the..." not 100% sure.